agent/src/main/java/com/microsoft/hydralab/agent/runner/maestro/MaestroResultReceiver.java

```diff
@@ -0,0 +1,96 @@
+import java.io.InputStreamReader;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * @author zhoule
+ * @date 07/11/2023
+ */
+
+public class MaestroResultReceiver extends Thread {
+    private final InputStream inputStream;
+    private final Logger logger;
+    private final MaestroListener listener;
+    private static final String KEY_CASE_NAME = "caseName";
+    private static final String KEY_TEST_SECONDS = "testSeconds";
+    private static final String KEY_ERROR = "error";
+    private boolean isTestRunFailed = false;
+
+    public MaestroResultReceiver(InputStream inputStream, MaestroListener listener, Logger logger) {
+        this.inputStream = inputStream;
+        this.logger = logger;
+        this.listener = listener;
+    }
+
+    public void run() {
+        listener.testRunStarted();
+        try {
+            InputStreamReader isr = new InputStreamReader(inputStream, "UTF-8");
+            BufferedReader bufferedReader = new BufferedReader(isr);
+            String line;
+            while ((line = bufferedReader.readLine()) != null) {
+                logger.info(line);
+                if (line.startsWith("[Passed]")) {
+                    Map<String, String> caseInfo = analysisCaseInfo(line);
+                    listener.testEnded(caseInfo.get(KEY_CASE_NAME), Integer.parseInt(caseInfo.get(KEY_TEST_SECONDS)));
+                } else if (line.startsWith("[Failed]")) {
+                    Map<String, String> caseInfo = analysisCaseInfo(line);
+                    listener.testFailed(caseInfo.get(KEY_CASE_NAME), Integer.parseInt(caseInfo.get(KEY_TEST_SECONDS)), caseInfo.get(KEY_ERROR));
+                } else if (line.contains("Debug output")) {
+                    isTestRunFailed = true;
+                    logger.info("Start to analysis debug output");
+                }
+                if (isTestRunFailed) {
+                    if (LogUtils.isLegalStr(line, Const.RegexString.LINUX_ABSOLUTE_PATH, false)
+                            || LogUtils.isLegalStr(line, Const.RegexString.WINDOWS_ABSOLUTE_PATH, false)) {
+                        listener.testRunFailed(line);
+                    }
+                }
+            }
+            listener.testRunEnded();
+            isr.close();
+            bufferedReader.close();
+        } catch (IOException e) {
+            logger.info("Exception:" + e);
+            e.printStackTrace();
+        } finally {
+            synchronized (this) {
+                notify();
+            }
+        }
+    }
+
+    private Map<String, String> analysisCaseInfo(String line) {
+        Map<String, String> infoMap = new HashMap<>();
+        String[] msg = line.split(" ");
+        if (msg.length < 3) {
+            infoMap.put(KEY_CASE_NAME, "caseParseError");
+            infoMap.put(KEY_TEST_SECONDS, "0");
+        }
+        infoMap.put(KEY_CASE_NAME, msg[1]);
+        String testSeconds = msg[2].replace("s", "").replace("(", "").replace(")", "");
+        try {
```

Comment:
It would be better if you could add a comment as the comparison between the Before and After of the string.